### ISSUE TITLE
Handling datatypes

### ISSUE THREAD

**Issue by mhsjlw** [mhsjlw/pocket-minecraft-protocol#1]
Here's a list of datatypes ProtoDef doesn't handle

> Note: We should also rename anything with `[]` to `_array` or something similar

These I suspect are doable:
- [ ] int
- [ ] long
- [ ] byte
- [ ] short
- [ ] byte[] 
- [ ] string
- [ ] float
- [ ] uuid
- [ ] vector3

These may take some more work and we have to take a look at MiNET
- [ ] offline_message_data_id
- [ ] ipendpoint
- [ ] ipendpoint[]
- [ ] item
- [ ] metadatadictionary
- [ ] entitylocations
- [ ] blockrecords
- [ ] records
- [ ] playerattributes
- [ ] playerrecords
- [ ] blockcoordinates
- [ ] entitymotions
- [ ] itemstacks
- [ ] metadataints
- [ ] recipes
- [ ] nbt

**Comment by rom1504** [comment]
Need much more details info.
From just these information, I see nothing that cannot be implemented by protodef.
The problems might be in the details.

For example what is "byte[]" ? is it a length-prefixed array like in mcpc ?

what is vector3 ?

Need info about what are these other complex datatypes.

Here are some example on how complex types can be described with protodef https://github.com/PrismarineJS/minecraft-data/blob/master/data/1.8/protocol.json#L2

some other examples there https://github.com/PrismarineJS/prismarine-nbt/blob/master/nbt.json

Basically we need to know what kind of types/fields are present in these complex datatypes.

**Comment by mhsjlw** [comment]
Ok, I'll start collecting information from MiNET and wiki.vg as those are relevant. Then I'll update the issue.

**Comment by mhsjlw** [comment]
This is what I've found so far
https://github.com/NiclasOlofsson/MiNET/blob/master/src/MiNET/MiNET/Net/MCPE%20Protocol%20Documentation.md this is pretty good for some datatypes 

https://github.com/NiclasOlofsson/MiNET/blob/master/src/MiNET/MiNET/Utils/Vector3.cs `vector3`
https://github.com/NiclasOlofsson/MiNET/blob/master/src/MiNET/MiNET/Utils/Nbt.cs `nbt`

OFFLINE_MESSAGE_DATA_ID is always 
```
byte[]
{ 0x00, 0xff, 0xff, 0x00, 0xfe, 0xfe, 0xfe, 0xfe, 0xfd, 0xfd, 0xfd, 0xfd, 0x12, 0x34, 0x56, 0x78 }
```

https://github.com/NiclasOlofsson/MiNET/blob/master/src/MiNET/MiNET/Net/UUID.cs#L57 `uuid` this seems to be the format

**Comment by mhsjlw** [comment]
Also, for 0.13.2 the current protocol version is 39

**Comment by mhsjlw** [comment]
Here are most of the datatypes: https://github.com/PocketMine/RakLib/blob/master/Binary.php (RakNet related only)
